---
output: 
    html_document:
        code_folding: hide
        df_print: tibble
        highlight: default
        theme: paper
        toc: true
        toc_depth: 5
        toc_float: true
        css: src/style/style.css

always_allow_html: yes
---

```{r setup, include=FALSE}
###############################################################################
## Recommended R-version                                                     ##

# module purge;source /camp/stp/babs/working/software/modulepath_new_software_tree_2018-08-13;module load pandoc/2.2.3.2-foss-2016b;ml R/4.0.3-foss-2020a


if (!require("remotes")){
  install.packages("remotes")
}


if (!require("renv")){
  remotes::install_github("rstudio/renv")
}


if (!file.exists("renv.lock")){
    renv::init()
} else {
    renv::restore(prompt = FALSE)
}


## Done                                                                      ##
###############################################################################


###############################################################################
## Set the environment                                                       ##

# if (!require("remotes")){
#   install.packages("remotes")
# }
# 
# remotes::install_github("rstudio/renv")

## Manual startup process
# cd into /analyses/Main_analysis
# start R
# do renv::activate()
# do renv::restore()
# Future: remove renv. files in main


if (!file.exists("renv.lock")){
    renv::init(
    )
} else {
    renv::restore(
        #project = "../../",
        prompt=FALSE
    )
}

projectDir <- gsub("scripts/bulkRNAseq_workflow/analyses/Main_Analysis", "", getwd())
workdir <- gsub("scripts/bulkRNAseq_workflow","workdir/",getwd())
figureCount <- 1
tableCount <- 1
upload.to.database <- TRUE

#renv::install("bioc::DESeq2")
#renv::install("bioc::clusterProfiler")
#renv::install("decusInLabore/biologicSeqTools2")
#renv::install("jokergoo/ComplexHeatmap")

## Done                                                                      ##
###############################################################################


###############################################################################
## Set knitr options                                                         ##

knitr::opts_chunk$set(
    tidy = F,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE,
    #dev="png"
    dev = "png", dev.args = list(type = "cairo-png"),
    fig.width=7,  # default setting is 7 
    fig.height=5,  # default setting is 5
    fig.retina=1  # default setting is 1
    #fig.path = "images/"
)

##                                                                           ##
###############################################################################
```


<!-- Essential: Load Obio Data Object ---------------------------------------->
```{r child = '01-A1.define.settings.Rmd', eval=TRUE}
```

<!-- Essential:Create Obio object     ---------------------------------------->
```{r child = '02-Create.bioLOGIC.object.Rmd', eval=TRUE}
```

```{r, echo=FALSE, eval=TRUE, warning=FALSE, results=F}
###############################################################################
## load biologic object                                                      ##

source("load.biologic.robj.R")

## Done                                                                      ##
###############################################################################

###############################################################################
## Set report variables                                                      ##

# Obio@parameterList$workdir - available 
Obio@parameterList[["figureCount"]] <- 1
Obio@parameterList[["tableCount"]] <- 1
Obio@parameterList[["upload.results.to.database"]] <- TRUE
Obio@parameterList[["shinyBaseServerURL"]] <- "shiny-bioinformatics.crick.ac.uk"
Obio@parameterList[["urlString"]] <- "biologic.crick.ac.uk"

## Done setting report variables                                             ##
###############################################################################

##############################################################################
# In this case - set project params                                         ##
# Obio@documentationParams <- list(
#     "title" = "Zdhcc20 Crispr Bulk RNA-Seq Analysis",
#     "subtitle" = "Tony Ocasio and Goran Tomic - Tate Lab",
#     "abstract" = "In this project Cripr knockouts for Zdhcc20 were analysed."
# )

## Done                                                                      ##
###############################################################################

###############################################################################
## Set password for database access                                          ##
###############################################################################
## Set database password                                                     ##
# Get database password from file

if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

## Loading the BABS password ##
if (Obio@parameterList[["upload.results.to.database"]]){
    FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
    dbTable <- read.delim(
      FN,
      header = F,
      sep = "\t",
      stringsAsFactors = F
    )
}

 Obio@dbDetailList[["db.pwd"]] <- as.vector(dbTable[1,1])
##                                                                           ##
###############################################################################
## Done                                                                      ##
###############################################################################

###############################################################################
## Save biologic object with additional variables                            ##

# source("save.biologic.robj.R")

## Done                                                                      ##
###############################################################################

```


<!-- Essential: Load Obio Data Object ---------------------------------------->
```{r child = '01-B.1.set.parameters.Rmd', eval=TRUE}
```

<!-- Optional set data access parameters for R-users ------------------------->
```{r child = '02-B.1.2.set.data.access.parameters.for.R.data.import.Rmd', eval=TRUE}
```

<!-- Optional: Retrieve reference gene sets----------------------------------->
```{r child = '03-B.2.retrieve.reference.gene.sets.Rmd', eval=TRUE}
```

<!-- Essential: Add RSEM count matrix and TPM matrix ------------------------->
```{r child = '04-A.1.add.data.to.analysis.object.Rmd', eval=TRUE}
```

<!-- Optional: Add alignment summary to report-------------------------------->
<!-- Modify - copy nf-core outputs            -------------------------------->
```{r child = '05-B.3.alignment.summary.Rmd', eval=TRUE}
```

<!-- Optional 4: (creates output only): Analysis Guide -->
```{r child = '06-B.4.analysis.guide.Rmd', eval=TRUE}
```

<!-- Optional 5: (creates output only): Display design table -->
```{r child = '07-B.5.display.design.table.and.formula.table.Rmd', eval=TRUE}
```

<!-- Essential: DEseq2 analysis -->
<!-- Creates PCA -->
```{r child = '08-B.6.DEseq2.PCA.Rmd', eval=TRUE}
```

<!-- Essential: DEseq2 analysis -->
<!-- Creates LRT and DGE analyses -->
```{r child = '09-B.7.DEseq2.LRT.DGE.Rmd', eval=TRUE}
```

<!-- Optional 8: (creates output only): Sample Characterization -->
<!-- Coefficient of variation and cluster dendrogram -->
```{r child = '10-B.8.sample.characterization.Rmd', eval=TRUE}
```

<!-- Optional 9: (output only): PCA sample characterization Chapter/Module -->
```{r child = '11-B.9.pca.characterization.Rmd', eval=TRUE}
```

<!-- Optional 10: (output only): Gene Category Heatmaps Chapter/Module -->
```{r child = '12-B.10.heatmaps.Rmd', eval=TRUE}
```

<!-- Optional 11: (output only): MA- and Volcanoplot Chapter/Module -->
```{r child = '13-B.11.volcanoplots.ma.plots.Rmd', eval=TRUE}
```

<!-- Optional 12: (output only): Category Enrichment Hypergeometric Test Chapter/Module -->
```{r child = '14-B.12.cat.enrichment.HG.Rmd', eval=TRUE}
```

<!-- Optional 13: (output only): Category Enrichment fGSEA Method Chapter/Module -->
```{r child = '15-B.13.cat.enrichment.fGSEA.Rmd', eval=TRUE}
```


<!-- Optional 12: (output only): LRT Analysis Chapter/Module -->
```{r child = '16-B.14.LRT.plots.Rmd', eval=TRUE}
```

<!-- Optional 13: (output only): Documentation Chapter/Module -->
```{r child = '17-B.15.bulkRNAseq.documentation.Rmd', eval=TRUE}
```

```{r, echo=FALSE, eval=TRUE, warning=FALSE, results=F}
###############################################################################
## Save Obio Object                                                          ##

## Remove password from file
Obio@dbDetailList[["db.pwd"]] <- NULL

source("save.biologic.robj.R")

##                                                                           ##
###############################################################################


```






```{r, echo=FALSE, eval=TRUE, warning=FALSE, results=F}

## Try to retrieve project data from db ##
db.pwd2 <- "zU3ufd9L"
db.user2 <- "reader"
host2 <- "clvd1-db-u-p-17.thecrick.org"
projectParams <- Obio@documentationParams

tryCatch({
    dbDB = DBI::dbConnect(
        drv = RMySQL::MySQL(), 
        user = db.user2, 
        password = db.pwd2, 
        host = host2, 
        dbname = "clarity_shadow"
    )
    dfProposal <-  DBI::dbGetQuery(
        dbDB, 
        paste0("SELECT * FROM clarify_asf_proposals WHERE project_name ='",Obio@projectDetailList$lims.id,"'")
    )
    dbDisconnect(dbDB)
}, error = function(x) {
    message("Project Database could not be reached or has no entry in Obio@parameterList$lims.id for this analysis.")
})

###############################################################################
## Helper
firstup <- function(x) {
    substr(x, 1, 1) <- toupper(substr(x, 1, 1))
    x
}
##
###############################################################################


if (exists("dfProposal")){
    if (!is.na(dfProposal[1,"ProjectAlias"]) & dfProposal[1,"ProjectAlias"] != ""){
        projectParams[["title"]] = paste0(dfProposal[1,"ProjectAlias"], " - ", dfProposal[1,"project_name"])
    }
    
    if (!is.na(dfProposal[1,"project_user"]) & dfProposal[1,"project_user"] != ""){
        labString <- firstup(dfProposal[1,"user_lab"])
        labString <- substr(labString, 1, (nchar(labString) - 1))
      
        projectParams[["subtitle"]] = paste0(labString, " Lab - ", dfProposal[1,"project_user"])
        projectParams[["subtitle"]] <- gsub("^ Lab - ", "", projectParams[["subtitle"]])
        
    }
    
    if (!is.na(dfProposal[1,"proposal_text"]) & dfProposal[1,"proposal_text"] != ""){
        projectParams[["abstract"]] = dfProposal[1,"proposal_text"]
       
        
    }
}
   
## Escape all special characters
projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("([.|()/\\^{}+$*?]|\\[|\\])", " ", x)
) 

projectParams <- lapply(
  projectParams, function(x) 
  #gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\1", x)
  gsub("\\\n", " ", x)
) 


```



## Documentation
```{r documentation, eval=TRUE, echo=F, results=T, error=F}
#renv::snapshot(prompt=FALSE)

print(paste0("Projectfolder: ", getwd()))
print(paste0("Project ID: ", Obio@parameterList$project_id))

sessionInfo()
```

---
title: "`r projectParams$title`"
subtitle:  "`r projectParams$subtitle`"
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    "`r projectParams$abstract`"


---
